File: 61.旋转链表/rotateRight.go
package problem0061

type ListNode struct {
	Val  int
	Next *ListNode
}

func rotateRight(head *ListNode, k int) *ListNode {
	if head == nil || head.Next == nil || k == 0 {
		return head
	}
	// 1. 寻找到最后一个节点以及链表长度
	last, length := head, 1
	for last.Next != nil {
		length++
		last = last.Next
	}
	// 2. 精简移动步数
	k %= length
	if k == 0 {
		return head
	}
	// 3. 成环
	last.Next = head
	// 4. 旋转链表 = 移动k步 = 倒数第k个节点做新表头(length-k), 倒数第k-1个节点做新的表尾
	start := head
	for i := 0; i < length-k-1; i++ {
		start = start.Next
	}
	result := start.Next
	start.Next = nil
	return result
}

File: 61.旋转链表/rotateRight.js
// 旋转链表
// 给你一个链表的头节点 head ，旋转链表，将链表每个节点向右移动 k 个位置。
// 0 <= 链表长度 <= 500
// 0 <= k <= 2 * 109
function rotateRight(head, k) {
	if (!head || !head.next || k === 0) {
		return head
	}
	// 由于存在 k 大于链表长度的情况，因此需要求出链表的长度，以及链表最后一个元素
	let length = 1
	let cur = head
	while (cur.next) {
		length++
		cur = cur.next
	}
	// 实际上只需要移动 k % length 步
	k %= length
	// 成环
	cur.next = head
	// 旋转链表 = 移动k步 = 倒数第k个节点做新表头(length-k), 倒数第k-1个节点做新的表尾
	cur = head
	for (let i = 0; i < length-k-1; i++) {
		cur = cur.next
	}
	// 此时cur是表尾
	const newHead = cur.next
	cur.next = null
	return newHead
}